sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -167,19 +167,14 @@ case class FileSourceScanExec(
       partitionSchema = relation.partitionSchema,
       relation.sparkSession.sessionState.conf)
 
+  private var fileListingTime = 0L
+
   @transient private lazy val selectedPartitions: Seq[PartitionDirectory] = {
     val optimizerMetadataTimeNs = relation.location.metadataOpsTimeNs.getOrElse(0L)
     val startTime = System.nanoTime()
     val ret = relation.location.listFiles(partitionFilters, dataFilters)
     val timeTakenMs = ((System.nanoTime() - startTime) + optimizerMetadataTimeNs) / 1000 / 1000
-
-    metrics("numFiles").add(ret.map(_.files.size.toLong).sum)
-    metrics("metadataTime").add(timeTakenMs)
-
-    val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
-    SQLMetrics.postDriverMetricUpdates(sparkContext, executionId,
-      metrics("numFiles") :: metrics("metadataTime") :: Nil)
-
```

Comment:
@xuanyuanking . Could you check which Spark versions have this bug? This code was introduced [at Spark 2.2.0](https://github.com/apache/spark/commit/60977889eaecdf28adc6164310eaa5afed488fa1).

Reply:
All versions after 2.2.0 have this bug, this PR fix the conner case(but all case by sql) in the commit. If we test by:
```
spark.read.parquet($filepath)
```
the metrics show correctly cause `relation.partitionSchemaOption` is None, `selectedPartitions` initializing not triggered at wrong place. But if we test as the added UT:
```
sql("select * from table partition=xx")
```
the metrics goes wrong.
I changed the JIRA description and title more detailed, please help me to check whether I explain clearly. Thanks :)